src/dune_engine/action_builder.ml

```diff
@@ -309,60 +313,59 @@ struct
     end
 
     module Poly_memo = Memo.Poly.Async (Function)
-
-    let eval_pred x = Fdecl.get Dep.eval_pred x
-
     open Memo.Build.O
 
-    let rec exec : type a. a t -> (a * Dep.Set.t) Memo.Build.t = function
-      | Pure x -> Memo.Build.return (x, Dep.Set.empty)
+    let merge_facts a b = Dep.Map.union a b ~f:(fun _ a _ -> Some a)
```

Comment:
Maybe the type `fact` should come with its own `merge` function. Then you won't have to duplicate `merge_facts` which is the same as `Facts.union` (the latter also has a useful `assert`).

Reply:
(This would make a collection of facts a "join semilattice", where there is the "empty" collection of facts, and a "join" operation that combines facts in a non-contradictory manner.)

Reply:
sounds good, done